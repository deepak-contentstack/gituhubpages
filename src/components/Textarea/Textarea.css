@import '../../../tokens/variables.scss';
@import '../../../tokens/tokens.css';
@import '../../static/styles/css/typography.css';
@import '../../static/styles/css/variable.css';
@import '../../static/styles/css/common.css';

.Textarea {
    display: flex;

    &__char-length{
        color: $color-font-base;
        float:right;
        font-size: $size-font-tiny;
        font-weight: $font-weight-regular;
        margin-right: -$space-15;
        margin-top: $space-5;
        visibility: hidden;
        &--reached{
            color: $color-base-orange-20;
        }
    }

    &--with-count{
        display: block;
        margin-bottom:-$space-2;
        padding-right: $space-20;
    }
}

.Textarea__textarea {
    background-color: $color-brand-white-base;
    border: 0;
    border-bottom: $space-2 solid transparent;
    box-sizing: content-box;
    color: $color-font-black;
    flex-basis: 100%;
    font-family: $font-family-primary;
    font-size:$size-font-large;
    line-height: $space-30;
    max-height: 31.25rem;
    min-height: $space-30;
    outline: none;
    padding: $space-4 $space-10 $space-4 $space-10;
    resize: vertical;
    
    &::-webkit-input-placeholder {
        color: $color-font-secondary;
    }
    &:hover{
        background-color: $color-brand-secondary-lighter;
        border-radius: $radii-8;
    }
    &:focus {
        background-color: $color-brand-white-base;
        border-radius: $radii-8;
        box-shadow: 0 $space-4 $space-24 rgba(0, 0, 0, 0.1);
        outline:none;

        & ~ .Textarea__char-length{
            visibility: visible;
        }
    }
}



.Textarea--small {
    width: 7.5rem;
}

.Textarea--medium {
    width: 15rem;
}

.Textarea--large {
    width: 26.25rem;
}

.Textarea--x-large {
    width: 48.13rem;
}

.Textarea--full {
    width: 100%;
}

.Textarea--resize {
    position: relative;
    .Textarea__textarea {
        height: $space-30;
        resize: none;
    }
    &--hover {
        height: $space-2;
        &:hover {
            .Textarea--resize__icon svg path {
                fill : $color-brand-primary-base;
            }
            .Textarea--resize__line {
                &>div{
                background-color: $color-brand-primary-base;
                border-radius: 0 0 $radii-8 $radii-8;
                height:$space-2;
        }
            }
        }
    }
    &__icon {
        bottom: 0;
        cursor: row-resize;
        padding:0 $space-5 $space-4 0;
        position: absolute;
        right: 0;
    }
    &__line { 
        bottom: 0;
        cursor: row-resize;
        height: $space-4;
        padding: 0 $space-4 ;
        position: absolute;
        width: 100%;
    }
}

.Textarea--disabled {
    & .Textarea__textarea {
        background-color: $color-brand-white-base;
        color:$color-base-gray-35;
        cursor: not-allowed;
    }
}

.Textarea--negative {
    & .Textarea__textarea {
        border-bottom: $space-2 solid $color-border-warning;
    }
}