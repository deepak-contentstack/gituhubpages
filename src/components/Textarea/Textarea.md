# Textarea

## Best Practises

- Use along with Field Label.
- Can be used where multiline text is required.
- Stretch the input area to view the long text.
