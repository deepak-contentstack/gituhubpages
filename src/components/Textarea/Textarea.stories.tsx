import React from 'react';
import Textarea from './Textarea';
import notes from './Textarea.md'
import { action } from '@storybook/addon-actions';
import DocSubData from '../../utils/components/DocSubData';

const widthList: any = ['small', 'medium', 'large', 'x-large', 'full'];
const componentDescription = 'The Textarea component allows the developer to render a text area for long text requirements.';
const usageDescription = 'Can be used as a text input area where text to be entered can be of multiple lines. Depending on the end goal of the requirement, the user is free to choose any Textarea component.';

export default {
  title: 'Components / Textarea',
  component: Textarea,
  argTypes: {
    name: {
      control: { type: 'text' },
      description: 'Specifies the name attribute of input element.',
      table: {
        type: {
          summary: 'string'
        }
      }
    },
    id: {
      control: { type: 'text' },
      description: 'Specifies the id attribute of input element.',
      table: {
        type: {
          summary: 'string'
        }
      }
    },
    placeholder: {
      control: { type: 'text' },
      description: 'Used to give a hint before entering value.',
      table: {
        type: {
          summary: 'string'
        },
        defaultValue: {
          summary: 'Type something...'
        }
      }
    },
    width: {
      control: { type: 'select', options: widthList },
      description: 'Sets the width of the text area .It can be `small`| `medium` |`large` |`x-large`| `full` .',
      table: {
        type: {
          summary: 'union'
        },
        defaultValue: {
          summary: 'full'
        },
      }
    },
    required: {
      control: { type: 'boolean' },
      description: 'Sets that the corresponding input is required and cannot be blank/empty.',
      table: {
        type: {
          summary: 'boolean'
        },
        defaultValue: {
          summary: false
        }
      }
    },
    disabled: {
      control: { type: 'boolean' },
      description: 'Used to disable the input field.',
      table: {
        type: {
          summary: 'boolean'
        },
        defaultValue: {
          summary: false
        }
      }
    },
    value: {
      control: { type: 'text' },
      description: 'Lets you add value in the text area.',
      table: {
        type: {
          summary: 'string'
        }
      }
    },
    rows: {
      control: { type: 'number' },
      description: 'Specifies the row space taken by the text area.',
      table: {
        type: {
          summary: 'number'
        }
      }
    },
    error: {
      control: { type: 'boolean' },
      description: 'Used to visually indicate an error.',
      table: {
        type: {
          summary: 'boolean'
        },
        defaultValue: {
          summary: false
        }
      }
    },
    showCharacterCount: {
      control: { type: 'boolean' },
      description: 'Shows the count of characters entered in the text area.',
      table: {
        type: {
          summary: 'boolean'
        },
        defaultValue: {
          summary: false
        }
      }
    },
    maxLength: {
      control: { type: 'number' },
      description: 'Used to limit the length of the input value characters.',
      table: {
        type: {
          summary: 'number'
        }
      },
    },
    onChange: {
      control: { type: 'disabled' },
      description: 'The event occurs when the value of an element has been changed.',
      table: {
        type: {
          summary: 'ChangeEventHandler'
        }
      }
    },
    inputRef: {
      control: { type: 'disabled' },
      description: 'This will add a reference to the text area.',
      table: {
        type: {
          summary: 'RefObject'
        }
      }
    },
    testId: {
      control: { type: 'text' },
      description: 'Used for testing purposes.',
      table: {
        type: {
          summary: 'text'
        },
        defaultValue: {
          summary: 'cs-text-area'
        }
      }
    },
    className: {
      control: { type: 'disabled' },
      description: 'Pass the class names to be appended to this prop.',
      table: {
        type: {
          summary: 'string'
        }
      }
    },
  },
  parameters: {
    componentSubtitle: <DocSubData componentDescription={componentDescription} usageDescription={usageDescription} />,
    notes: { notes },
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/QHcXcnGhGDBQ49J1kRwzYf/02-%E2%80%A2-Molecules?node-id=1234%3A12512'
    },
  }
}
const tableDisable = { table: { disable: true } }

export const Default = (args) => {
  return (
    <Textarea
      className={args.className}
      name={args.name}
      id={args.id}
      error={args.error}
      required={args.required}
      width={args.width}
      disabled={args.disabled}
      value={args.value}
      placeholder={args.placeholder}
      rows={args.rows}
      maxLength={args.maxLength}
      showCharacterCount={args.showCharacterCount}
      onChange={action('onChange')}
    ></Textarea>
  )
}
Default.args = {
  name: "someInput",
  id: "someInput",
  rows: 2,
}
Default.argTypes = {
  inputRef: tableDisable,
  testId: tableDisable,
}

export const Disabled = (args) => {
  return (<Textarea
    className={args.className}
    name={args.name}
    id={args.id}
    disabled={args.disabled}
    value={args.value}
  ></Textarea>
  )
}
Disabled.args = {
  name: "someInput",
  id: "someInput",
  value: "Some description",
  disabled: "true",
}
Disabled.argTypes = {
  className: tableDisable,
  placeholder: tableDisable,
  error: tableDisable,
  required: tableDisable,
  width: tableDisable,
  maxLength: tableDisable,
  showCharacterCount: tableDisable,
  rows: tableDisable,
  inputRef: tableDisable,
  testId: tableDisable,
  onChange: tableDisable,
}
Disabled.parameters = {
  docs: {
    storyDescription: 'Can be used where we need to make an input area disabled.',
  },
}

export const WithError = (args) => {
  return (<Textarea
    name={args.name}
    id={args.id}
    error={args.error}
    width={args.width}
    value={args.value}
    onChange={action('onChange')}
  ></Textarea>
  )
}
WithError.args = {
  name: "someInput",
  id: "someInput",
  error: 'true',
}
WithError.argTypes = {
  className: tableDisable,
  placeholder: tableDisable,
  disabled: tableDisable,
  required: tableDisable,
  maxLength: tableDisable,
  showCharacterCount: tableDisable,
  rows: tableDisable,
  inputRef: tableDisable,
  testId: tableDisable,
}
WithError.parameters = {
  docs: {
    storyDescription: 'Can be used to indicate error virtually.',
  },
}

export const MediumSize = (args) => {
  return (
    <Textarea
      className={args.className}
      name={args.name}
      id={args.id}
      width={args.width}
      value={args.value}
      rows={args.rows}
    ></Textarea>
  )
}
MediumSize.args = {
  name: "someInput",
  id: "someInput",
  width: 'medium'
}
MediumSize.argTypes = {
  className: tableDisable,
  placeholder: tableDisable,
  disabled: tableDisable,
  required: tableDisable,
  error: tableDisable,
  maxLength: tableDisable,
  showCharacterCount: tableDisable,
  inputRef: tableDisable,
  testId: tableDisable,
  onChange: tableDisable,
}
MediumSize.parameters = {
  docs: {
    storyDescription: 'Can set width of text area as small, medium, large, x-large or full.',
  },
}

export const WithCharCount = (args) => {
  return (<Textarea
    name={args.name}
    id={args.id}
    width={args.width}
    placeholder={args.placeholder}
    showCharacterCount={args.showCharacterCount}
    maxLength={50}
    onChange={action('onChange')}
  ></Textarea>
  )
}
WithCharCount.args = {
  name: "someInput",
  id: "someInput",
  width: 'medium',
  showCharacterCount: 'true',
  maxLength: '50',
}
WithCharCount.argTypes = {
  className: tableDisable,
  value: tableDisable,
  disabled: tableDisable,
  required: tableDisable,
  error: tableDisable,
  rows: tableDisable,
  inputRef: tableDisable,
  testId: tableDisable,
}
WithCharCount.parameters = {
  docs: {
    storyDescription: 'Can be used to show the count of characters entered in the text area.',
  },
}


