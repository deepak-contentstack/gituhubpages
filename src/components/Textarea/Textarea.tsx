import React, {
  Component,
  ChangeEventHandler
} from 'react';
import cn from 'classnames';
import './Textarea.css';
import withDeprecatedProp from '../../utils/hooks/depricatedPropsHoc';

export type TextareaProps = {
  /**
   *Specifies the name attribute of input element.
   */
  name?: string;
  /**
   *Specifies the id attribute of input element.
   */
  id?: string;
  /**
   *Used to give a hint before entering value.
   */
  placeholder?: string;
  /**
   *Pass the class names to be appended to this prop.
   */
  className?: string;
  /**
   *Sets the width of the text area .It can be `small`| `medium` |`large` |`x-large`| `full` .
   */
  width?: 'small' | 'medium' | 'large' | 'x-large' | 'full';
  /**
   *Used to limit the length of the input value characters.
   */
  maxLength?: number;
  /**
   *Sets that the corresponding input is required and cannot be blank/empty.
   */
  required?: boolean;
  /**
   *Used to disable the input field.
   */
  disabled?: boolean;
  /**
   *Lets you add value in the text area.
   */
  value?: string;
  /**
   *Specifies the row space taken by the text area.
   */
  rows?: number;
  /**
   *Used to visually indicate an error.
   */
  error?: boolean;
  /**
   *Shows the count of characters entered in the text area.
   */
  showCharacterCount?: boolean;
  /**
   *The event occurs when the value of an element has been changed.
   */
  onChange?: ChangeEventHandler<HTMLTextAreaElement>;
  /**
   *This will add a reference to the text area.
   */
  inputRef?: React.Ref<HTMLTextAreaElement>
  /**
   *Used for testing purposes.
   */
  testId?: string;
  /**
   * Provides the directionality of the text.
   */
  textDirection?: 'ltr' | 'rtl' | 'auto'
} & typeof defaultProps;

export interface TextareaState {
  value?: string;
}

const defaultProps = {
  disabled: false,
  required: false,
  width: 'full',
  placeholder: 'Type something...',
  testId: 'cs-text-area',
  dir: 'auto'
};
export class Textarea extends Component<TextareaProps, TextareaState> {
  static defaultProps = defaultProps;

  state = {
    value: this.props.value,
  };

  UNSAFE_componentWillReceiveProps(nextProps: TextareaProps) {
    if (this.props.value !== nextProps.value) {
      this.setState({
        value: nextProps.value,
      });
    }
  }


  render() {
    const {
      className,
      placeholder,
      maxLength,
      onChange,
      disabled,
      required,
      error,
      width,
      value,
      name,
      rows,
      id,
      inputRef,
      showCharacterCount,
      testId,
      textDirection,
      ...otherProps
    } = this.props;

    const widthClass = `Textarea--${width}`;
    const classNames = cn('Textarea', className, [widthClass], {
      'Textarea--disabled': disabled,
      'Textarea--negative': error,
      'Textarea--with-count': showCharacterCount,
    });


    return (
      <div className={classNames} data-test-id={testId}>
        <textarea
          aria-label={name}
          dir={textDirection}
          className="Textarea__textarea"
          id={id}
          rows={rows}
          disabled={disabled}
          onChange={e => {
            if (onChange) {
              onChange(e);
            }
            this.setState({ value: e.target.value });
          }}
          placeholder={disabled && !this.state.value ? '-' : placeholder}
          name={name}
          maxLength={maxLength}
          value={disabled ? value : this.state && this.state.value}
          ref={inputRef}
          {...otherProps}
        />
        {showCharacterCount && <span className={`Textarea__char-length${this.state.value && (this.state.value.length === maxLength) ? ' Textarea__char-length--reached' : ''}`}>{`${(this.state.value && this.state.value.length ? this.state.value.length : 0)} / ${maxLength}`}</span>}
      </div>
    );
  }
}

export default withDeprecatedProp(Textarea, { 'showCharCount': 'showCharacterCount' });