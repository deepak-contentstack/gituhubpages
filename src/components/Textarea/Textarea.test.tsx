import React from 'react'
import renderer from 'react-test-renderer'
import Textarea from './Textarea'
import { mount, shallow } from 'enzyme';
import EnzymeToJson from 'enzyme-to-json';
describe('Textarea', () => {
  it('Textarea with name', () => {
    const TextareaName = () => {
      return (
        <Textarea name="inputText" />
      )
    }
    const tree = renderer.create(<TextareaName />).toJSON()
    expect(tree).toMatchSnapshot()
  }),

    it('Textarea with id', () => {
      const TextareaId = () => {
        return (
          <Textarea id="inputText" />
        )
      }
      const tree = renderer.create(<TextareaId />).toJSON()
      expect(tree).toMatchSnapshot()
    }),

    it('Textarea with placeholder', () => {
      const element = mount(<Textarea placeholder="Enter Something..." />);
      expect(element.find('textarea').props().placeholder).toBe('Enter Something...');
    }),

    it('Textarea with classname', () => {
      const element = mount(<Textarea testId="custom-test-id" className="test-classname" placeholder="Enter Something..." />);
      expect(element.find('[data-test-id="custom-test-id"]').props().className.includes('test-classname')).toBe(true);
    }),

    it('render Textarea with small width', () => {
      const element = mount(<Textarea width="small" />);
      expect(element.find('[data-test-id="cs-text-area"]').props().className.includes('small')).toBe(true);
    }),

    it('render Textarea with large width', () => {
      const element = mount(<Textarea width="large" />);
      expect(element.find('[data-test-id="cs-text-area"]').props().className.includes('large')).toBe(true);
    }),

    it('render Textarea with medium width', () => {
      const element = mount(<Textarea width="medium" />);
      expect(element.find('[data-test-id="cs-text-area"]').props().className.includes('medium')).toBe(true);
    }),

    it('render Textarea with x-large width', () => {
      const element = mount(<Textarea width="x-large" />);
      expect(element.find('[data-test-id="cs-text-area"]').props().className.includes('x-large')).toBe(true);
    }),

    it('render Textarea with full width', () => {
      const element = mount(<Textarea width="full" />);
      expect(element.find('[data-test-id="cs-text-area"]').props().className.includes('full')).toBe(true);
    }),

    it('Textarea with max-length', () => {
      const TextareaMaxLength = () => {
        return (
          <Textarea maxLength={50} />
        )
      }
      const tree = renderer.create(<TextareaMaxLength />).toJSON()
      expect(tree).toMatchSnapshot()
    }),

    it('Textarea with required', () => {
      const TextareaRequired = () => {
        return (
          <Textarea required={true} />
        )
      }
      const tree = renderer.create(<TextareaRequired />).toJSON()
      expect(tree).toMatchSnapshot()
    }),

    it('Textarea with disabled', () => {
      const TextareaDisabled = () => {
        return (
          <Textarea disabled={true} />
        )
      }
      const tree = renderer.create(<TextareaDisabled />).toJSON()
      expect(tree).toMatchSnapshot()
    }),

    it('Textarea with value', () => {
      const TextareaValue = () => {
        return (
          <Textarea value="Some text" />
        )
      }
      const tree = renderer.create(<TextareaValue />).toJSON()
      expect(tree).toMatchSnapshot()
    }),

    it('Textarea with rows', () => {
      const TextareaRows = () => {
        return (
          <Textarea rows={2} />
        )
      }
      const tree = renderer.create(<TextareaRows />).toJSON()
      expect(tree).toMatchSnapshot()
    }),

    it('Textarea with error', () => {
      const TextareaError = () => {
        return (
          <Textarea error={true} />
        )
      }
      const tree = renderer.create(<TextareaError />).toJSON()
      expect(tree).toMatchSnapshot()
    }),

    it('Textarea with showCharacterCount', () => {
      const TextareaShowCharacterCount = () => {
        return (
          <Textarea showCharacterCount={true} />
        )
      }
      const tree = renderer.create(<TextareaShowCharacterCount />).toJSON()
      expect(tree).toMatchSnapshot()
    }),

    it('Call onChange method of Textarea', () => {
      const mockOnChange = jest.fn();
      const TextareaOnChange = ({ onChangeAction }) => {
        return (
          <Textarea onChange={onChangeAction}></Textarea>
        )
      };
      const element = mount(<TextareaOnChange onChangeAction={mockOnChange} />);
      element.find('textarea').simulate('change');
      expect(mockOnChange).toBeCalledTimes(1);
    }),

    it('componentWillRecieveProps', () => {
      const element = shallow(<Textarea value="initial" />);
      expect(element.props().value).toEqual('initial');
      element.setProps({ value: 'next' });
      expect(element.props().value).toEqual('next');
    });

  it('Textarea with dir as auto', () => {
    const element = mount(<Textarea dir="auto" placeholder="Enter Something..." />);
    expect(element.find('textarea').props().dir).toBe('auto')
  })
  it('Textarea with dir as rtl', () => {
    const element = mount(<Textarea dir="rtl" placeholder="Enter Something..." />);
    expect(element.find('textarea').props().dir).toBe('rtl')
  })
  it('Textarea with dir as ltr', () => {
    const element = mount(<Textarea dir="ltr" placeholder="Enter Something..." />);
    expect(element.find('textarea').props().dir).toBe('ltr')
  })
})
