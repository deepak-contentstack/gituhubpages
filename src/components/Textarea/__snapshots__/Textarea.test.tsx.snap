// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Textarea with disabled 1`] = `
<div
  className="Textarea Textarea--full Textarea--disabled"
  data-test-id="cs-text-area"
>
  <textarea
    className="Textarea__textarea"
    dir="auto"
    disabled={true}
    onChange={[Function]}
    placeholder="-"
  />
</div>
`;

exports[`Textarea Textarea with error 1`] = `
<div
  className="Textarea Textarea--full Textarea--negative"
  data-test-id="cs-text-area"
>
  <textarea
    className="Textarea__textarea"
    dir="auto"
    disabled={false}
    onChange={[Function]}
    placeholder="Type something..."
  />
</div>
`;

exports[`Textarea Textarea with id 1`] = `
<div
  className="Textarea Textarea--full"
  data-test-id="cs-text-area"
>
  <textarea
    className="Textarea__textarea"
    dir="auto"
    disabled={false}
    id="inputText"
    onChange={[Function]}
    placeholder="Type something..."
  />
</div>
`;

exports[`Textarea Textarea with max-length 1`] = `
<div
  className="Textarea Textarea--full"
  data-test-id="cs-text-area"
>
  <textarea
    className="Textarea__textarea"
    dir="auto"
    disabled={false}
    maxLength={50}
    onChange={[Function]}
    placeholder="Type something..."
  />
</div>
`;

exports[`Textarea Textarea with name 1`] = `
<div
  className="Textarea Textarea--full"
  data-test-id="cs-text-area"
>
  <textarea
    aria-label="inputText"
    className="Textarea__textarea"
    dir="auto"
    disabled={false}
    name="inputText"
    onChange={[Function]}
    placeholder="Type something..."
  />
</div>
`;

exports[`Textarea Textarea with required 1`] = `
<div
  className="Textarea Textarea--full"
  data-test-id="cs-text-area"
>
  <textarea
    className="Textarea__textarea"
    dir="auto"
    disabled={false}
    onChange={[Function]}
    placeholder="Type something..."
  />
</div>
`;

exports[`Textarea Textarea with rows 1`] = `
<div
  className="Textarea Textarea--full"
  data-test-id="cs-text-area"
>
  <textarea
    className="Textarea__textarea"
    dir="auto"
    disabled={false}
    onChange={[Function]}
    placeholder="Type something..."
    rows={2}
  />
</div>
`;

exports[`Textarea Textarea with showCharacterCount 1`] = `
<div
  className="Textarea Textarea--full Textarea--with-count"
  data-test-id="cs-text-area"
>
  <textarea
    className="Textarea__textarea"
    dir="auto"
    disabled={false}
    onChange={[Function]}
    placeholder="Type something..."
  />
  <span
    className="Textarea__char-length"
  >
    0 / undefined
  </span>
</div>
`;

exports[`Textarea Textarea with value 1`] = `
<div
  className="Textarea Textarea--full"
  data-test-id="cs-text-area"
>
  <textarea
    className="Textarea__textarea"
    dir="auto"
    disabled={false}
    onChange={[Function]}
    placeholder="Type something..."
    value="Some text"
  />
</div>
`;
